Client/Audio/PortAudio.cpp

```
@@ -0,0 +1,196 @@
+}
+
+void Babel::PortAudio::setSampleRate(int32_t nb)
+{
+    this->_samplerate = nb;
+}
+
+int32_t Babel::PortAudio::getSampleRate() const
+{
+    return (this->_samplerate);
+}
+
+void Babel::PortAudio::setRecordTime(int32_t nb)
+{
+    this->_recordtime = nb;
+}
+
+int32_t Babel::PortAudio::getRecordTime() const
+{
+    return (this->_recordtime);
+}
+
+void Babel::PortAudio::openStream()
+{
+    int32_t err = paNoError;
+    PaStreamParameters inputParameters;
+    PaStreamParameters outputParameters;
+
+    inputParameters.device = Pa_GetDefaultInputDevice();
+    if (inputParameters.device == paNoDevice) {
+        std::cerr << "No imput devices found" << std::endl;
+        throw std::exception();
+    }
+    inputParameters.channelCount = _number_channels;
+    inputParameters.sampleFormat = PA_SAMPLE_TYPE;
+    inputParameters.suggestedLatency = Pa_GetDeviceInfo(inputParameters.device)->defaultLowInputLatency;
+    inputParameters.hostApiSpecificStreamInfo = NULL;
+    outputParameters.device = Pa_GetDefaultOutputDevice();
+    if (outputParameters.device == paNoDevice) {
+        std::cerr << "No default output device found" << std::endl;
+        throw std::exception();
+    }
+    outputParameters.channelCount = _number_channels;
+    outputParameters.sampleFormat =  PA_SAMPLE_TYPE;
+    outputParameters.suggestedLatency = Pa_GetDeviceInfo(outputParameters.device)->defaultLowOutputLatency;
+    outputParameters.hostApiSpecificStreamInfo = NULL;
+    err = Pa_OpenStream(&_stream, &inputParameters, &outputParameters, _samplerate, _frames_per_buffer, paClipOff, NULL, NULL);
+    if (err != paNoError || !_stream) {
+        _stream = NULL;
+        std::cerr << "Could not open output stream" << std::endl;
+        throw std::exception();
+    }
+}
+
+void Babel::PortAudio::startStream()
+{
+    int32_t err = paNoError;
+
+    if (!_stream) {
+        std::cerr << "Audio stream not set" << std::endl;
+        throw std::exception();
```

please throw an exception derived from BabelException